pkg/cmd/flags.go

```diff
@@ -0,0 +1,172 @@
+// identify if a given service version needs to be cloned.
+type OptionalAutoClone struct {
+	OptionalBool
+}
+
+// Parse returns a service version.
+//
+// The returned version is either the same as the input argument `v` or it's a
+// cloned version if the input argument was either active or locked.
+func (ac *OptionalAutoClone) Parse(v *fastly.Version, sid string, c api.Interface) (*fastly.Version, error) {
+	// if user didn't provide --autoclone flag
+	if !ac.Value && (v.Active || v.Locked) {
+		return nil, fmt.Errorf("service version %d is not editable", v.Number)
+	}
+	if ac.Value && (v.Active || v.Locked) {
+		version, err := c.CloneVersion(&fastly.CloneVersionInput{
+			ServiceID:      sid,
+			ServiceVersion: v.Number,
+		})
+		if err != nil {
+			return nil, fmt.Errorf("error cloning service version: %w", err)
+		}
+		return version, nil
+	}
+
+	// Treat the function as a no-op if the version is editable.
+	return v, nil
+}
+
+// getLatestActiveVersion returns the latest active service version.
+func getLatestActiveVersion(vs []*fastly.Version) (*fastly.Version, error) {
+	for _, v := range vs {
+		if v.Active {
+			return v, nil
+		}
+	}
+	return nil, fmt.Errorf("error locating latest active service version")
+}
+
+// getLatestEditableVersion returns the latest editable service version.
+//
+// The returned version will be ahead of the latest active version. If there is
+// no editable version above the latest active, an error will be returned.
+//
+// When no --version flag is provided, this algorithm helps handle cases where
+// a command (such as `backend create`) is executable multiple times, as the
+// latest editable version will be reused and subsequently will contain each
+// new backend created from the prior executions.
+//
+// There could be a scenario where a customer has a single service version
+// which is activated and so there would be no match for an editable version
+// without us first cloning it. The act of cloning should be a decision made by
+// the user (i.e. --autoclone) and so this function will return an error if no
+// editable version available.
+func getLatestEditableVersion(vs []*fastly.Version) (*fastly.Version, error) {
+	for _, v := range vs {
+		if v.Active {
+			break
+		}
+		if !v.Active && !v.Locked {
+			return v, nil
+		}
+	}
+	return nil, fsterr.RemediationError{
+		Inner:       errors.New("error retrieving an editable service version: no editable version found"),
+		Remediation: "Please retry with --version=[latest|active] and --autoclone",
+	}
+}
+
+// getSpecifiedVersion returns the specified service version.
+func getSpecifiedVersion(vs []*fastly.Version, version string) (*fastly.Version, error) {
+	i, err := strconv.Atoi(version)
+	if err != nil {
+		return nil, err
+	}
+
+	for _, v := range vs {
+		if v.Number == i {
+			return v, nil
+		}
+	}
+
+	return nil, fmt.Errorf("error getting specified service version: %s", version)
```

Comment:
I'd probably reword this to make it clear that the version was not found as opposed to some other error.